Change "message" in Get-Credential to mandatory

iSazonov opened the pull request:
Revert change from PR #1904 
1. Set "message" to optional is not best practice, from https://msdn.microsoft.com/en-us/library/dd878348%28v=vs.85%29.aspx :
   •Each parameter set must have at least one unique parameter. If possible, make this parameter a mandatory parameter.
   Set both "message" and "title" to optional open a way to "anonymous" request of credential. That is bad. This provokes writing bad code.
2. Set "message" to optional also breaks backward compatibility: script running in version 6 will not work in the previous versions if "message" is empty.

iSazonov committed:
Change "message" to mandatory

Close #2306

iSazonov committed:
Merge pull request #1 from iSazonov/getcredential

Change "message" in Get-Credential to mandatory

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

iSazonov commented:
Hmm.. I don't understand why tests failed. Please help me.

iSazonov committed:
Change get-credential tests

Change test to check mandatory message parameter

iSazonov committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into getcredential

iSazonov committed:
Fix get-credential tests

Check exception
ParameterArgumentValidationError,Microsoft.PowerShell.Commands.GetCredentialCommand

mirichmo commented:
@lzybkr Please take a look. I think you have been dealing with the iterations of this issue.

lzybkr commented:
I tested these changes in an interactive session and it does not fix #2306, so it seems we assumed incorrectly that making `-Message` optional introduced that problem.

Did you want to investigate a fix for #2306 before we accept this PR?

lzybkr commented:
Never mind about fixing #2306 - the fix is/will be in #2330.

iSazonov commented:
Clear about #2306. I removed Close #2306.

> Did you want to investigate a fix for #2306 before we accept this PR?

Yes, because in #2330 already added test for "It "Get-Credential without parameters" { " - it is unnecessary work.

iSazonov commented:
@lzybkr  I can confirm that changing "message" to mandatory fix #2306!
1. compile FullCRL with mandatory=true (my fix) - no second prompt
2. compile FullCRL  with mandatory=false (w/o my fix) - second prompt appear

lzybkr commented:
I tried the changes made by @daxian-dbw - I think it provides exactly the experience we want without making `-Message` mandatory.

Having a mandatory parameter in a parameter set is a guideline, not a requirement. In this specific case, there are benefits to not making `-Message` mandatory - one can pass a custom `-Title` and use the default *localized* message.

And correct me if I'm wrong, but are there really compatibility issues?  All previous versions of PowerShell support calling `Get-Credential` with no parameters, #2330 just improves the user experience.

iSazonov commented:
I revoke "compatibility" statement. In previous versions, an user only receives one more question, but a script will work.

> Having a mandatory parameter in a parameter set is a guideline, not a requirement

Yes, it is not a requirement but the guide says about best practice for script writer that comes from the best user experience.

But!  I was very conservative. I modeled various combinations of parameters (default and non-default, localized and non-localized) and think that as best practice the script writers will have to specify the message and title in a pair without the requirement that message is mandatory.

So you can dicard the PR. @lzybkr  thanks for great discussion!

lzybkr commented:
Thanks @iSazonov - I really appreciate your attention to details like this, hopefully we'll see more contributions from you in the future.
